future.go

```diff
@@ -129,38 +129,41 @@ func NewFuture() (fut *Future) {
 // AppendPush appends the push response to the future.
 // Note: it works only before SetResponse() or SetError()
 func (fut *Future) AppendPush(resp *Response) {
+	fut.mutex.Lock()
+	defer fut.mutex.Unlock()
+
 	if fut.isDone() {
 		return
 	}
```

Comment:
Is it possible to have a race condition under the fut.done channel? If not, maybe it is worth to move lock lines like this?
```suggestion
	if fut.isDone() {
		return
	}
	fut.mutex.Lock()
	defer fut.mutex.Unlock()
```

Reply:
You are right, there is no problem with fut.done, but it will be possible an error with fut.ready:

thread1:
1. isDone() == false
-------------------<<< thread2 calls SetResponse() or SetError(), it closes fut.ready
3. fut.mutex.Lock()
4. fut.pushes = append(fut.pushes, resp)
5. fut.ready <- struct{}{}
6. fut.mutex.Unlock()

As a result it will be an error on step 5 - writting to the closed channel fut.ready.

Reply:
Thank you for explanation!